Fix: Respect parent properties in case of merged schemas by allOf

musaprg opened the pull request:
fixes #697 

Currently, oapi-codegen ignores original (parent) fields if `allOf` is defined. OpenAPI schema doesn't seem to refuse a combination of general fields (`properties`) and `allOf`.

For example, the following schema

```
    PersonWithMorePropertiesOutsideOfAllOf:
      type: object
      description: |
        This is a person record as returned from a Create endpoint. It contains
        all the fields of a Person, with an additional property outside of allOf directives.
      properties:
        additionalProperty:
          type: string
      required: [ additionalProperty ]
      allOf:
        - $ref: "#/components/schemas/Person"
```

will be converted into the following type definition, which seems to be wrong result.

```
type PersonWithMorePropertiesOutsideOfAllOf = Person
```

The expected result would be like:

```
type PersonWithMorePropertiesOutsideOfAllOf struct {
	// Embedded struct due to allOf(#/components/schemas/Person)
	Person `yaml:",inline"`
	// Embedded fields due to inline allOf schema
	AdditionalProperty string `json:"additionalProperty"`
}
```

The known approach is _normalization_ as proposed in other similar project like `REFACTOR_ALLOF_WITH_PROPERTIES_ONLY` in `opneapi-generator`. This PR is inspired by the idea.

> `REFACTOR_ALLOF_WITH_PROPERTIES_ONLY`: When set to true, refactor schema with allOf and properties in the same level to a schema with allOf only and, the allOf contains a new schema containing the properties in the top level.
https://github.com/OpenAPITools/openapi-generator/blob/8b5b5a74c333b809c5a651366656257ec8a6fef3/docs/customization.md?plain=1#L567

---

@oapi-codegen/maintainers edit to PR message:
Closes #697 
Closes #931
Closes #1710
Closes #1960
Closes #2102 
Partially improves ##1139
Supersedes #717

musaprg committed:
Respect parent properties in merging allOf schemas

Signed-off-by: Kotaro Inoue <k.musaino@gmail.com>

musaprg committed:
Add test case to check if parent propeties are included in the merged schema

Signed-off-by: Kotaro Inoue <k.musaino@gmail.com>

musaprg committed:
Make generate

Signed-off-by: Kotaro Inoue <k.musaino@gmail.com>

musaprg commented:
@jamietanna Would you take a look at this PR? Thanks.

tymofij commented:
Thank you for this, I hope it gets reviewed soon

mromaszewicz committed:
Merge remote-tracking branch 'upstream/main' into pr/1415

# Conflicts:
#	internal/test/all_of/v1/openapi.gen.go
#	internal/test/all_of/v2/openapi.gen.go

mromaszewicz commented:
Thanks for this PR. I'm patching it up, and it actually closes quite a few related bugs. I will push a couple of fix-up commits to your branch.

mromaszewicz committed:
fix(codegen): only inject parent siblings into allOf merge when structural

The unconditional injection of the parent (with AllOf cleared) as a
final allOf member correctly merges structural siblings like
properties/required, but it also forces n>=2 in mergeSchemas for pure
wrappers like `{allOf: [$ref: X]}`, which previously took the n==1
fast-path and returned the referenced type. Without the fast-path
those wrappers regenerate as inlined anonymous structs and lose their
named-type identity (regression visible in
examples/output-options/preferskipoptionalpointer).

Add hasStructuralSiblings() so injection runs only when the parent has
Properties, Required, or AdditionalProperties. Pure wrappers keep the
fast-path. Also copy the parent's Description into the merged result —
the merge does not carry Description, so it was being silently dropped.

Addresses #697, #931, #1710, #1960, #2102.
Partial fix for #1139 (anonymous-schema hoisting requires further work).

mromaszewicz commented:
@greptileai

mromaszewicz commented on internal/test/issues/issue-1087/deps/deps.gen.go:
```diff
@@ -31,7 +31,22 @@ type BaseError struct {
 }
 
 // Error defines model for Error.
-type Error = BaseError
+type Error struct {
+	// Code The underlying http status code
+	Code int32 `json:"code"`
+
+	// Domain The domain where the error is originating from as defined by the service
+	Domain string `json:"domain"`
+
+	// Message A simple message in english describing the error and can be returned to the consumer
+	Message string `json:"message"`
+
+	// Metadata Any additional details to be conveyed as determined by the service. If present, will return map of key value pairs
+	Metadata *map[string]string `json:"metadata,omitempty"`
+
+	// Reason A reason code specific to the service and can be used to identify the exact issue. Should be unique within a domain
+	Reason string `json:"reason"`
+}
```
This will be fixed by putting it behind a flag. This breaking behavior is intentional, because the old code was incorrect. We'll default to the new approach and allow users to opt out.

mromaszewicz commented on pkg/codegen/schema.go:
```diff
@@ -1012,3 +1035,22 @@ func combinedSchemaExtensions(r *openapi3.SchemaRef) map[string]any {
+// are also excluded for now: their strict-equality check in
+// mergeOpenapiSchemas conflates the bool zero value with "unset" and would
+// regress simple wrappers like {allOf: [X-with-nullable:true]}.
+func hasStructuralSiblings(s *openapi3.Schema) bool {
+	if s == nil {
+		return false
+	}
+	return len(s.Properties) > 0 ||
+		len(s.Required) > 0 ||
+		s.AdditionalProperties.Has != nil ||
+		s.AdditionalProperties.Schema != nil
+}
```
@greptileai, you are mistaken here. This PR fixes many issues, and this is an intentional change to fix #931. `required` on the parent is intended to change a field's structure, in our case, dropping the indirect since it becomes required.

mromaszewicz committed:
fix(codegen): gate allOf sibling merge behind compatibility flag

Address Greptile P1 review on the previous commit:

1. append(schema.AllOf, ...) could mutate kin-openapi's backing array
   if the parsed slice carried spare capacity. Allocate a fresh slice
   with make() before appending.

2. Merging the parent's structural siblings into an allOf result is a
   breaking change to generated Go: a schema like
   {allOf: [Base], properties: {extra}} previously produced
   `type X = Base` (interchangeable with Base) and now produces a
   distinct struct (not interchangeable). Per project practice
   (cf. OldMergeSchemas, OldEnumConflicts, OldAliasing) gate behavior
   changes behind a compatibility flag. Add
   Compatibility.OldAllOfSiblingMerging (yaml: old-allof-sibling-merging,
   default false). Update configuration-schema.json with a matching
   description so users hit by downstream breakage can find the knob.

The Description copy is gated on the same flag for consistency.

Greptile also suggested narrowing hasStructuralSiblings to drop the
pure-required case. Keeping Required in the trigger because that is the
exact scenario issue #931 reports: a parent that adds `required: [foo]`
to an allOf chain currently produces a Go type whose `Foo` field is
`*string` with omitempty, ignoring the spec's required constraint.
Required is structural in OAS terms — it changes field type in the
generated struct — and the compatibility flag is the safety net for
users whose downstream code depended on the alias output.

mromaszewicz commented:
@greptileai comments addressed, review again, make sure to look at latest commit.
